sdk/src/time_stamp.rs

```diff
@@ -448,273 +404,277 @@ fn get_validator_type(sig_alg: &bcder::Oid, hash_alg: &bcder::Oid) -> Option<Str
                     continue;
                 }
-            };
+            }
+        }
 
-            let hash_alg = &signer_info.digest_algorithm.algorithm;
-            let sig_alg = &signer_info.signature_algorithm.algorithm;
-
-            // grab signing certificate
-            let sig_val = &signer_info.signature;
-            let mut signing_key_der = Vec::<u8>::new();
-            cert.tbs_certificate
-                .subject_public_key_info
-                .encode_ref()
-                .write_encoded(bcder::Mode::Der, &mut signing_key_der)?;
-
-            // verify signature of timestamp signature
-            let validated_res: Result<bool> = if _sync {
-                #[cfg(feature = "openssl")]
-                {
-                    let validator = get_local_validator(sig_alg, hash_alg)?;
-                    validator.validate(&sig_val.to_bytes(), &tbs, &signing_key_der)
-                }
+        // build CMS structure to verify
+        let tbs = match signer_info.signed_attributes_digested_content() {
+            Ok(sdc) => match sdc {
+                Some(tbs) => tbs,
+                None => match &sd.content_info.content {
+                    Some(d) => d.to_bytes().to_vec(),
+                    None => return Err(Error::CoseTimeStampMismatch),
+                },
+            },
+            Err(_) => {
+                last_err = Error::CoseTimeStampMismatch;
+                continue;
+            }
+        };
 
-                #[cfg(not(feature = "openssl"))]
-                {
-                    Ok(false)
-                }
-            } else {
-                #[cfg(feature = "openssl")]
-                {
-                    let validator = get_local_validator(sig_alg, hash_alg)?;
-                    validator.validate(&sig_val.to_bytes(), &tbs, &signing_key_der)
-                }
+        let hash_alg = &signer_info.digest_algorithm.algorithm;
+        let sig_alg = &signer_info.signature_algorithm.algorithm;
 
-                #[cfg(target_arch = "wasm32")]
-                {
+        // grab signing certificate
+        let sig_val = &signer_info.signature;
+        let mut signing_key_der = Vec::<u8>::new();
+        cert.tbs_certificate
+            .subject_public_key_info
+            .encode_ref()
+            .write_encoded(bcder::Mode::Der, &mut signing_key_der)?;
+
+        // Verify signature of timestamp signature.
+        #[cfg(feature = "openssl")]
+        validate_timestamp_sig(sig_alg, hash_alg, sig_val, &tbs, &signing_key_der)?;
+
+        #[cfg(not(feature = "openssl"))]
+        {
+            #[cfg(target_arch = "wasm32")]
+            {
+                if _sync {
+                    validate_timestamp_sig(sig_alg, hash_alg, sig_val, &tbs, &signing_key_der)?;
+                } else {
+                    // TO REVIEW: Worth keeping this WASM-specific async path alive, or can we fully switch over to the synchronous path? (I'd prefer the latter.)
```

Comment:
Reviewers: Please let me know if the WASM-specific async path is worth preserving. I'd love to simplify down to the new sync-only path.